core/src/core/transaction.rs

```diff
@@ -971,24 +982,40 @@ impl Transaction {
 /// and outputs.
 pub fn cut_through(inputs: &mut Vec<Input>, outputs: &mut Vec<Output>) -> Result<(), Error> {
 	// assemble output commitments set, checking they're all unique
-	let mut out_set = HashSet::new();
-	let all_uniq = { outputs.iter().all(|o| out_set.insert(o.commitment())) };
-	if !all_uniq {
+	outputs.sort_unstable_by(|v1, v2| v1.commitment().0.cmp(&v2.commitment().0));
+	if outputs
+		.windows(2)
+		.any(|pair| pair[0].commitment().0[..] == pair[1].commitment().0[..])
+	{
 		return Err(Error::AggregationError);
 	}
-
-	let in_set = inputs
-		.iter()
-		.map(|inp| inp.commitment())
-		.collect::<HashSet<_>>();
-
-	let to_cut_through = in_set.intersection(&out_set).collect::<HashSet<_>>();
-
-	// filter and sort
-	inputs.retain(|inp| !to_cut_through.contains(&inp.commitment()));
-	outputs.retain(|out| !to_cut_through.contains(&out.commitment()));
-	inputs.sort();
-	outputs.sort();
+	inputs.sort_unstable_by(|v1, v2| v1.commitment().0.cmp(&v2.commitment().0));
+	let mut inps = (0..inputs.len()).peekable();
+	let mut inps_insert_idx = 0..;
+	let mut outs = (0..outputs.len()).peekable();
+	let mut outs_insert_idx = 0..;
+
+	let mut ncut = 0;
+	while let (Some(i), Some(i2)) = (inps.peek(), outs.peek()) {
```

Comment:
I've read through this a few times an I still don't fully understand what we're doing here...

Lets say - 
* inputs: `[A,B,F,H]`
* outputs: `[B,C,H,J]`

After cut-through we should see - 
* inputs: `[A,F]`
* outputs: `[C,J]`
* (removed: `[B,H]`)

Reading through the impl here - 
* We sort inputs and outputs (and check outputs are unique)
* then we iterate over peekable inputs and outputs
    * if they are equal we `next()` both and increment `ncut`
    * if not equal we `next()` one or the other 
* Then we `drain` slices of inputs and outputs using `ncut`

Given outputs `[B,C,H,J]` how do end up with `[C,J]` using a single drain with a single slice using ncut?
What does `ncut` do here? I'm not following this (and missing something obvious).

Reply:
Yup! Let's go through step by step:
^ means the _insert_idx will next yield this location,
* means the inps/outs will next yield this location
```
Step 1:
[^*A,B,F,H]
[^*B,C,H,J]
ncut = 0
A < B

Step 2:
[A,^*B,F,H]
[^*B,C,H,J]
ncut = 0
B == B

Step 3:
[A,^B,*F,H]
[^B,*C,H,J]
ncut = 1
F > C

Step 4:
[A,^B,*F,H]
[C,^C,*H,J]
ncut = 1
F < H


Step 5:
[A,F,^F,*H]
[C,^C,*H,J]
ncut = 1
H == H


Step 6:
[A,F,^F,H] *
[C,^C,H,*J]
ncut = 2
vec 1 has None to compare

Step 7:
vec1.drain(^..^+ncut)
[A,F,_,_]
[A,F]
vec2.drain(^..^+ncut)
[C,_,_,*J]
[C,J]
```

I think I wrote this out correctly... if there are any errors, let me know.

Essentially what we are doing is walking both lists in sorted order with four fingers, and backshifting anything that is not in the other list, which we are guaranteed via the sort order. When we finish, the drain of that particular range cuts out only the ones we've already inserted into the list earlier. It leaves (and then backshifts) the range after which we've yet to look at (but don't need to, because it is greater than the greatest element in the other list).

Technically I think you can compute ncut as (*-^), but I think there is some edge condition, so tracking ncut seemed more straightforward.

Reply:
:+1: Thanks for the explanation. Looks good.